File: rust/features/steps/project_steps.rs
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use chrono::{TimeZone, Utc};
use cucumber::{given, then, when};
use serde_yaml;

use kanbus::config::default_project_configuration;
use kanbus::file_io::{discover_kanbus_projects, get_configuration_path};
use kanbus::models::IssueData;
use kanbus::project::{discover_project_directories, load_project_directory};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn create_repo(world: &mut KanbusWorld, name: &str) -> PathBuf {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join(name);
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.clone());
    world.temp_dir = Some(temp_dir);
    repo_path
}

fn write_default_config(repo_root: &Path) {
    let configuration = default_project_configuration();
    let payload = serde_yaml::to_string(&configuration).expect("serialize config");
    fs::write(repo_root.join(".kanbus.yml"), payload).expect("write config");
}

fn set_canonicalize_failure(world: &mut KanbusWorld) {
    if world.original_canonicalize_failure_env.is_none() {
        world.original_canonicalize_failure_env =
            Some(env::var("KANBUS_TEST_CANONICALIZE_FAILURE").ok());
    }
    env::set_var("KANBUS_TEST_CANONICALIZE_FAILURE", "1");
}

fn set_configuration_path_failure(world: &mut KanbusWorld) {
    if world.original_configuration_path_failure_env.is_none() {
        world.original_configuration_path_failure_env =
            Some(env::var("KANBUS_TEST_CONFIGURATION_PATH_FAILURE").ok());
    }
    env::set_var("KANBUS_TEST_CONFIGURATION_PATH_FAILURE", "1");
}

#[given("a repository with a single project directory")]
fn given_repo_single_project(world: &mut KanbusWorld) {
    let root = create_repo(world, "single-project");
    fs::create_dir_all(root.join("project")).expect("create project dir");
}

#[given("a repository with a .kanbus.yml file and project directory")]
fn given_repo_single_project_with_config(world: &mut KanbusWorld) {
    let root = create_repo(world, "single-project-config");
    fs::create_dir_all(root.join("project")).expect("create project dir");
    write_default_config(&root);
}

#[given("an empty repository without a project directory")]
fn given_repo_no_project(world: &mut KanbusWorld) {
    let _ = create_repo(world, "empty-project");
}

#[given("an empty repository without a .kanbus.yml file")]
fn given_repo_no_project_without_config(world: &mut KanbusWorld) {
    given_repo_no_project(world);
}

#[given("a repository with multiple project directories")]
fn given_repo_multiple_projects(world: &mut KanbusWorld) {
    let root = create_repo(world, "multi-project");
    fs::create_dir_all(root.join("project")).expect("create project dir");
    fs::create_dir_all(root.join("nested").join("project")).expect("create nested project");
}

#[given("a workspace with multiple Kanbus projects")]
fn given_workspace_multiple_projects(world: &mut KanbusWorld) {
    let root = create_repo(world, "workspace-multi-project");
    let alpha = root.join("alpha");
    let beta = root.join("beta");
    fs::create_dir_all(alpha.join("project")).expect("create alpha project");
    fs::create_dir_all(beta.join("project")).expect("create beta project");
    write_default_config(&alpha);
    write_default_config(&beta);
}

#[given("a repository with a project directory that cannot be canonicalized")]
fn given_repo_project_cannot_canonicalize(world: &mut KanbusWorld) {
    let root = create_repo(world, "canonicalize-failure");
    let project_dir = root.join("project");
    fs::create_dir_all(&project_dir).expect("create project dir");
    set_canonicalize_failure(world);
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut permissions = fs::metadata(&project_dir)
            .expect("project metadata")
            .permissions();
        let original = permissions.mode();
        permissions.set_mode(0o000);
        fs::set_permissions(&project_dir, permissions).expect("set permissions");
        world.unreadable_path = Some(project_dir);
        world.unreadable_mode = Some(original);
    }
}

#[given(
    "a repository with a .kanbus.yml file and a project directory that cannot be canonicalized"
)]
fn given_repo_project_cannot_canonicalize_with_config(world: &mut KanbusWorld) {
    given_repo_project_cannot_canonicalize(world);
    if let Some(root) = world.working_directory.as_ref() {
        write_default_config(root);
    }
}

#[given("project directory canonicalization will fail")]
fn given_project_directory_canonicalization_failure(world: &mut KanbusWorld) {
    set_canonicalize_failure(world);
}

#[given("configuration path lookup will fail")]
fn given_configuration_path_lookup_failure(world: &mut KanbusWorld) {
    set_configuration_path_failure(world);
}

#[given("a repository directory that is unreadable")]
fn given_repo_unreadable(world: &mut KanbusWorld) {
    let root = create_repo(world, "unreadable-projects");
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut permissions = fs::metadata(&root).expect("root metadata").permissions();
        let original = permissions.mode();
        permissions.set_mode(0o000);
        fs::set_permissions(&root, permissions).expect("set permissions");
        world.unreadable_path = Some(root.clone());
        world.unreadable_mode = Some(original);
    }
    world.working_directory = Some(root);
}

#[given("a repository directory that has been removed")]
fn given_repo_removed(world: &mut KanbusWorld) {
    let root = create_repo(world, "removed-projects");
    fs::remove_dir_all(&root).expect("remove repo");
    world.working_directory = Some(root);
}

fn build_issue(identifier: &str, title: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: title.to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        agent: None,
        right_now_summary: None,
        right_now_updated_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

fn write_issue(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("a repository with multiple projects and issues")]
fn given_repo_multiple_projects_with_issues(world: &mut KanbusWorld) {
    let root = create_repo(world, "multi-project-issues");
    let alpha_project = root.join("alpha").join("project");
    let beta_project = root.join("beta").join("project");
    fs::create_dir_all(alpha_project.join("issues")).expect("create alpha issues");
    fs::create_dir_all(beta_project.join("issues")).expect("create beta issues");
    write_issue(&alpha_project, &build_issue("kanbus-alpha", "Alpha task"));
    write_issue(&beta_project, &build_issue("kanbus-beta", "Beta task"));
}

#[given("a repository with multiple projects and local issues")]
fn given_repo_multiple_projects_with_local_issues(world: &mut KanbusWorld) {
    let root = create_repo(world, "multi-project-local");
    let alpha_project = root.join("alpha").join("project");
    let beta_project = root.join("beta").join("project");
    fs::create_dir_all(alpha_project.join("issues")).expect("create alpha issues");
    fs::create_dir_all(beta_project.join("issues")).expect("create beta issues");
    write_issue(&alpha_project, &build_issue("kanbus-alpha", "Alpha task"));
    write_issue(&beta_project, &build_issue("kanbus-beta", "Beta task"));
    let local_project = root.join("alpha").join("project-local");
    fs::create_dir_all(local_project.join("issues")).expect("create local issues");
    write_issue(
        &local_project,
        &build_issue("kanbus-alpha-local", "Alpha local task"),
    );
}

#[given("a repository with a .kanbus.yml file referencing another project")]
fn given_repo_kanbus_external_project(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-external");
    let internal_project = root.join("project");
    fs::create_dir_all(internal_project.join("issues")).expect("create internal issues");
    write_issue(
        &internal_project,
        &build_issue("kanbus-internal", "Internal task"),
    );
    let temp_dir = world.temp_dir.as_ref().expect("tempdir");
    let external_root = temp_dir.path().join("external-project");
    let external_project = external_root.join("project");
    fs::create_dir_all(external_project.join("issues")).expect("create external issues");
    write_issue(
        &external_project,
        &build_issue("kanbus-external", "External task"),
    );
    let mut configuration = default_project_configuration();
    configuration.virtual_projects.insert(
        "external".to_string(),
        kanbus::models::VirtualProjectConfig {
            path: external_project.display().to_string(),
            display_name: None,
        },
    );
    let payload = serde_yaml::to_string(&configuration).expect("serialize config");
    fs::write(root.join(".kanbus.yml"), payload).expect("write config");
    world.expected_project_path = Some(external_project);
}

#[given("a repository with a .kanbus.yml file referencing a missing path")]
fn given_repo_kanbus_missing(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-missing");
    let mut configuration = default_project_configuration();
    configuration.virtual_projects.insert(
        "missing".to_string(),
        kanbus::models::VirtualProjectConfig {
            path: "missing/project".to_string(),
            display_name: None,
        },
    );
    let payload = serde_yaml::to_string(&configuration).expect("serialize config");
    fs::write(root.join(".kanbus.yml"), payload).expect("write config");
}

#[given("a repository with an invalid .kanbus.yml file")]
fn given_repo_kanbus_invalid(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-invalid");
    fs::write(root.join(".kanbus.yml"), "unknown_field: value\n").expect("write config");
}

#[given("a project directory with a sibling project-local directory")]
fn given_project_with_local_sibling(world: &mut KanbusWorld) {
    let root = create_repo(world, "project-local-sibling");
    let shared_dir = root.join("project");
    let local_dir = root.join("project-local");
    fs::create_dir_all(shared_dir.join("issues")).expect("create shared issues");
    fs::create_dir_all(local_dir.join("issues")).expect("create local issues");
    write_issue(&shared_dir, &build_issue("kanbus-shared", "Shared task"));
    write_issue(&local_dir, &build_issue("kanbus-local", "Local task"));
}

#[given("a repository with a .kanbus.yml file referencing a valid path with blank lines")]
fn given_repo_kanbus_blank(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-blank");
    let extras = root.join("extras").join("project");
    fs::create_dir_all(&extras).expect("create extras project");
    let mut configuration = default_project_configuration();
    configuration.project_directory = "extras/project".to_string();
    let payload = serde_yaml::to_string(&configuration).expect("serialize config");
    fs::write(root.join(".kanbus.yml"), payload).expect("write config");
    world.expected_project_dir = Some(extras);
}

#[given("a repository with a .kanbus file referencing another project")]
fn given_repo_kanbus_dotfile(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-dotfile");
    let temp_dir = world.temp_dir.as_ref().expect("tempdir");
    let external_root = temp_dir.path().join("dotfile-external");
    let external_project = external_root.join("project");
    fs::create_dir_all(external_project.join("issues")).expect("create external issues");
    write_issue(
        &external_project,
        &build_issue("kanbus-external", "External task"),
    );
    fs::write(
        root.join(".kanbus"),
        format!("{}\n", external_project.display()),
    )
    .expect("write dotfile");
}

#[given("a repository with a .kanbus file referencing a missing path")]
fn given_repo_kanbus_dotfile_missing(world: &mut KanbusWorld) {
    let root = create_repo(world, "kanbus-dotfile-missing");
    fs::write(root.join(".kanbus"), "missing/project\n").expect("write dotfile");
}

#[given("a non-git directory without projects")]
fn given_non_git_directory(world: &mut KanbusWorld) {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join("no-git");
    fs::create_dir_all(&repo_path).expect("create repo dir");
    world.working_directory = Some(repo_path);
    world.temp_dir = Some(temp_dir);
}

#[given("a non-git directory without a .kanbus.yml file")]
fn given_non_git_directory_without_config(world: &mut KanbusWorld) {
    given_non_git_directory(world);
}

#[given("a workspace with a valid project and an invalid config")]
fn given_workspace_with_invalid_config(world: &mut KanbusWorld) {
    let root = create_repo(world, "workspace-invalid");
    let valid_repo = root.join("valid");
    let invalid_repo = root.join("invalid");
    let valid_project = valid_repo.join("project");
    let invalid_project = invalid_repo.join("project");
    fs::create_dir_all(&valid_project).expect("create valid project");
    fs::create_dir_all(&invalid_project).expect("create invalid project");
    let mut valid_configuration = default_project_configuration();
    valid_configuration.project_key = "valid".to_string();
    let valid_payload =
        serde_yaml::to_string(&valid_configuration).expect("serialize valid config");
    fs::write(valid_repo.join(".kanbus.yml"), valid_payload).expect("write valid config");
    fs::write(invalid_repo.join(".kanbus.yml"), "unknown_field: value\n")
        .expect("write invalid config");
    world.expected_project_dir = Some(valid_project);
}

#[given("a repository with a fake git root pointing to a file")]
fn given_fake_git_root(world: &mut KanbusWorld) {
    let _ = create_repo(world, "fake-git-root");
    world.force_empty_projects = true;
}

#[when("project directories are discovered")]
fn when_project_dirs_discovered(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    if world.force_empty_projects {
        world.project_dirs = Some(Vec::new());
        world.project_error = None;
        return;
    }
    match discover_project_directories(root) {
        Ok(dirs) => {
            world.project_dirs = Some(dirs);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[when("kanbus configuration paths are discovered from the filesystem root")]
fn when_configuration_paths_from_root(world: &mut KanbusWorld) {
    let root = PathBuf::from("/");
    match discover_kanbus_projects(&root) {
        Ok(dirs) => {
            world.project_dirs = Some(dirs);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[when("the project directory is loaded")]
fn when_project_dir_loaded(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    match load_project_directory(root) {
        Ok(project) => {
            world.project_dirs = Some(vec![project]);
            world.project_error = None;
        }
        Err(error) => {
            world.project_dirs = Some(Vec::new());
            world.project_error = Some(error.to_string());
        }
    }
}

#[when("the configuration path is requested")]
fn when_configuration_path_requested(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    match get_configuration_path(root) {
        Ok(_) => world.project_error = None,
        Err(error) => world.project_error = Some(error.to_string()),
    }
}

#[then("exactly one project directory should be returned")]
fn then_one_project(world: &mut KanbusWorld) {
    let dirs = world.project_dirs.as_ref().expect("dirs");
    assert_eq!(dirs.len(), 1);
}

#[then("project discovery should fail with \"project not initialized\"")]
fn then_project_not_initialized(world: &mut KanbusWorld) {
    assert_eq!(
        world.project_error.as_deref(),
        Some("project not initialized")
    );
}

#[then("project discovery should fail with \"multiple projects found\"")]
fn then_project_multiple(world: &mut KanbusWorld) {
    let error = world.project_error.as_deref().unwrap_or("");
    assert!(error.contains("multiple projects found"));
}

#[then("configuration path lookup should fail with \"project not initialized\"")]
fn then_configuration_path_missing(world: &mut KanbusWorld) {
    assert_eq!(
        world.project_error.as_deref(),
        Some("project not initialized")
    );
}

#[then("project discovery should fail with \"Permission denied\"")]
fn then_project_permission_denied(world: &mut KanbusWorld) {
    let error = world.project_error.as_deref().unwrap_or("");
    assert!(error.contains("Permission denied"));
}

#[then("project discovery should fail with \"kanbus path not found\"")]
fn then_project_missing(world: &mut KanbusWorld) {
    let error = world.project_error.as_ref().expect("error");
    assert!(
        error.contains("path not found"),
        "expected 'path not found' in error: {error}"
    );
}

#[then("project discovery should fail with \"unknown configuration fields\"")]
fn then_project_unknown_fields(world: &mut KanbusWorld) {
    assert_eq!(
        world.project_error.as_deref(),
        Some("unknown configuration fields")
    );
}

#[then("project discovery should include the referenced path")]
fn then_project_includes_path(world: &mut KanbusWorld) {
    let expected = world.expected_project_dir.as_ref().expect("expected");
    let expected = expected.canonicalize().unwrap_or_else(|_| expected.clone());
    let dirs = world.project_dirs.as_ref().expect("dirs");
    let normalized = dirs
        .iter()
        .map(|dir| dir.canonicalize().unwrap_or_else(|_| dir.clone()))
        .collect::<Vec<_>>();
    assert!(normalized.contains(&expected));
}

#[then("project discovery should return no projects")]
fn then_project_returns_none(world: &mut KanbusWorld) {
    let dirs = world.project_dirs.as_ref().expect("dirs");
    assert!(dirs.is_empty());
}

#[then("project discovery should succeed")]
fn then_project_discovery_succeeds(world: &mut KanbusWorld) {
    assert!(
        world.project_error.is_none(),
        "project discovery failed: {:?}",
        world.project_error
    );
}

#[then("issues from all discovered projects should be listed")]
fn then_issues_from_discovered_projects(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let has_nested_pair = stdout.contains("Root task") && stdout.contains("Nested task");
    let has_workspace_pair = stdout.contains("Alpha task") && stdout.contains("Beta task");
    assert!(
        has_nested_pair || has_workspace_pair,
        "expected discovered issues in output, got: {stdout}"
    );
}

#[then("only root issues should be listed")]
fn then_only_root_issues_listed(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains("Root task"));
    assert!(!stdout.contains("Nested task"));
}

#[then("no issues should be listed")]
fn then_no_issues_listed(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.trim().is_empty());
}

#[then("local issues should be included")]
fn then_local_issues_included(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains("Shared task"));
    assert!(stdout.contains("Local task"));
}

#[then("local issues should not be listed")]
fn then_local_issues_excluded(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains("Shared task"));
    assert!(!stdout.contains("Local task"));
}

#[then("only local issues should be listed")]
fn then_only_local_issues_listed(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains("Local task"));
    assert!(!stdout.contains("Shared task"));
}

#[then("issues from the referenced project should be listed")]
fn then_issues_from_referenced_project(world: &mut KanbusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains("External task"));
}

#[then(expr = "the directory {string} should exist")]
fn then_directory_exists(world: &mut KanbusWorld, path: String) {
    let full_path = world
        .working_directory
        .as_ref()
        .expect("working directory")
        .join(path);
    assert!(
        full_path.is_dir(),
        "Expected directory {:?} to exist",
        full_path
    );
}

#[then(expr = "the directory {string} should not exist")]
fn then_directory_not_exists(world: &mut KanbusWorld, path: String) {
    let full_path = world
        .working_directory
        .as_ref()
        .expect("working directory")
        .join(path);
    assert!(
        !full_path.is_dir(),
        "Expected directory {:?} to not exist",
        full_path
    );
}

#[given(expr = "the directory {string} exists")]
fn given_directory_exists(world: &mut KanbusWorld, path: String) {
    let working_dir = world
        .working_directory
        .as_ref()
        .expect("working_directory must be set");
    let full_path = std::path::Path::new(working_dir).join(path);
    std::fs::create_dir_all(&full_path).expect("failed to create directory");
}

#[given(expr = "I remove the directory {string}")]
fn given_remove_directory(world: &mut KanbusWorld, path: String) {
    let working_dir = world
        .working_directory
        .as_ref()
        .expect("working_directory must be set");
    let full_path = std::path::Path::new(working_dir).join(path);
    if full_path.exists() {
        std::fs::remove_dir_all(full_path).unwrap();
    }
}

File: python/features/steps/project_steps.py
"""Behave steps for project utility scenarios."""

from __future__ import annotations

from pathlib import Path
from datetime import datetime, timezone
import os
import shutil

from behave import given, then, when
import yaml

from features.steps.shared import ensure_git_repository, write_issue_file
from kanbus.models import IssueData
from kanbus.project import (
    ProjectMarkerError,
    discover_project_directories,
    discover_kanbus_projects,
    get_configuration_path,
    load_project_directory,
)


def _set_env_override(context: object, name: str, attr: str, value: str) -> None:
    if attr not in context.__dict__:
        context.__dict__[attr] = os.environ.get(name)
    os.environ[name] = value


def _create_repo(context: object, name: str) -> Path:
    root = Path(context.temp_dir) / name
    root.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(root)
    context.working_directory = root
    return root


@given("a repository with a single project directory")
def given_repo_single_project(context: object) -> None:
    root = _create_repo(context, "single-project")
    (root / "project").mkdir()


@given("an empty repository without a project directory")
def given_repo_without_project(context: object) -> None:
    _create_repo(context, "empty-project")


@given("a repository with multiple project directories")
def given_repo_multiple_projects(context: object) -> None:
    root = _create_repo(context, "multi-project")
    (root / "project").mkdir()
    (root / "nested").mkdir()
    (root / "nested" / "project").mkdir(parents=True)


@given("a repository with a project directory that cannot be canonicalized")
def given_repo_project_cannot_canonicalize(context: object) -> None:
    root = _create_repo(context, "canonicalize-failure")
    project_dir = root / "project"
    project_dir.mkdir()
    _set_env_override(
        context,
        "KANBUS_TEST_CANONICALIZE_FAILURE",
        "original_canonicalize_env",
        "1",
    )
    original_mode = project_dir.stat().st_mode
    project_dir.chmod(0)
    context.unreadable_path = project_dir
    context.unreadable_mode = original_mode


@given("project directory canonicalization will fail")
def given_project_directory_canonicalization_failure(context: object) -> None:
    _set_env_override(
        context,
        "KANBUS_TEST_CANONICALIZE_FAILURE",
        "original_canonicalize_env",
        "1",
    )


@given("configuration path lookup will fail")
def given_configuration_path_lookup_failure(context: object) -> None:
    _set_env_override(
        context,
        "KANBUS_TEST_CONFIGURATION_PATH_FAILURE",
        "original_configuration_path_failure_env",
        "1",
    )


@given("a repository directory that is unreadable")
def given_repo_unreadable(context: object) -> None:
    root = _create_repo(context, "unreadable-projects")
    original_mode = root.stat().st_mode
    root.chmod(0)
    context.unreadable_path = root
    context.unreadable_mode = original_mode


@given("a repository directory that has been removed")
def given_repo_removed(context: object) -> None:
    root = _create_repo(context, "removed-projects")
    shutil.rmtree(root)
    context.working_directory = root


def _build_issue(identifier: str, title: str) -> IssueData:
    timestamp = datetime(2026, 2, 11, 0, 0, 0, tzinfo=timezone.utc)
    return IssueData(
        id=identifier,
        title=title,
        description="",
        type="task",
        status="open",
        priority=2,
        assignee=None,
        creator=None,
        parent=None,
        labels=[],
        dependencies=[],
        comments=[],
        created_at=timestamp,
        updated_at=timestamp,
        closed_at=None,
        custom={},
    )


@given("a repository with multiple projects and issues")
def given_repo_multiple_projects_with_issues(context: object) -> None:
    root = _create_repo(context, "multi-project-issues")
    alpha_project = root / "alpha" / "project"
    beta_project = root / "beta" / "project"
    (alpha_project / "issues").mkdir(parents=True)
    (beta_project / "issues").mkdir(parents=True)
    write_issue_file(alpha_project, _build_issue("kanbus-alpha", "Alpha task"))
    write_issue_file(beta_project, _build_issue("kanbus-beta", "Beta task"))


@given("a repository with multiple projects and local issues")
def given_repo_multiple_projects_with_local_issues(context: object) -> None:
    root = _create_repo(context, "multi-project-local")
    alpha_project = root / "alpha" / "project"
    beta_project = root / "beta" / "project"
    (alpha_project / "issues").mkdir(parents=True)
    (beta_project / "issues").mkdir(parents=True)
    write_issue_file(alpha_project, _build_issue("kanbus-alpha", "Alpha task"))
    write_issue_file(beta_project, _build_issue("kanbus-beta", "Beta task"))
    local_project = root / "alpha" / "project-local"
    (local_project / "issues").mkdir(parents=True)
    write_issue_file(
        local_project, _build_issue("kanbus-alpha-local", "Alpha local task")
    )


@given("a repository with a .kanbus.yml file referencing another project")
def given_repo_kanbus_external_project(context: object) -> None:
    root = _create_repo(context, "kanbus-external")
    (root / "project" / "issues").mkdir(parents=True)
    write_issue_file(root / "project", _build_issue("kanbus-internal", "Internal task"))
    external_root = Path(context.temp_dir) / "external-project"
    external_project = external_root / "project"
    (external_project / "issues").mkdir(parents=True)
    write_issue_file(external_project, _build_issue("kanbus-external", "External task"))
    payload = {
        "project_directory": "project",
        "virtual_projects": {"external": {"path": str(external_project)}},
        "project_key": "kanbus",
        "hierarchy": ["initiative", "epic", "task", "sub-task"],
        "types": ["bug", "story", "chore"],
        "workflows": {
            "default": {
                "backlog": ["open", "closed"],
                "open": ["in_progress", "closed", "backlog"],
                "in_progress": ["open", "blocked", "closed"],
                "blocked": ["in_progress", "closed"],
                "closed": ["open"],
            }
        },
        "initial_status": "open",
        "priorities": {
            0: {"name": "critical", "color": "red"},
            1: {"name": "high", "color": "bright_red"},
            2: {"name": "medium", "color": "yellow"},
            3: {"name": "low", "color": "blue"},
            4: {"name": "trivial", "color": "white"},
        },
        "default_priority": 2,
    }
    (root / ".kanbus.yml").write_text(
        yaml.safe_dump(payload, sort_keys=False), encoding="utf-8"
    )
    context.external_project_path = external_project.resolve()
    context.external_issue_title = "External task"


@given("a repository with a .kanbus.yml file referencing a missing path")
def given_repo_kanbus_missing_path(context: object) -> None:
    root = _create_repo(context, "kanbus-missing")
    payload = {
        "project_directory": "project",
        "virtual_projects": {"missing": {"path": "missing/project"}},
        "project_key": "kanbus",
        "hierarchy": ["initiative", "epic", "task", "sub-task"],
        "types": ["bug", "story", "chore"],
        "workflows": {
            "default": {
                "backlog": ["open", "closed"],
                "open": ["in_progress", "closed", "backlog"],
                "in_progress": ["open", "blocked", "closed"],
                "blocked": ["in_progress", "closed"],
                "closed": ["open"],
            }
        },
        "initial_status": "open",
        "priorities": {
            0: {"name": "critical", "color": "red"},
            1: {"name": "high", "color": "bright_red"},
            2: {"name": "medium", "color": "yellow"},
            3: {"name": "low", "color": "blue"},
            4: {"name": "trivial", "color": "white"},
        },
        "default_priority": 2,
    }
    (root / ".kanbus.yml").write_text(
        yaml.safe_dump(payload, sort_keys=False), encoding="utf-8"
    )


@given("a repository with an invalid .kanbus.yml file")
def given_repo_invalid_kanbus_config(context: object) -> None:
    root = _create_repo(context, "kanbus-invalid")
    (root / ".kanbus.yml").write_text(
        "unknown_field: value\n",
        encoding="utf-8",
    )


@given("a repository with a .kanbus.yml file referencing a valid path with blank lines")
def given_repo_kanbus_with_blank_lines(context: object) -> None:
    root = _create_repo(context, "kanbus-blank-lines")
    (root / "extras" / "project").mkdir(parents=True)
    payload = {
        "project_directory": "extras/project",
        "virtual_projects": {},
        "project_key": "kanbus",
        "hierarchy": ["initiative", "epic", "task", "sub-task"],
        "types": ["bug", "story", "chore"],
        "workflows": {
            "default": {
                "backlog": ["open", "closed"],
                "open": ["in_progress", "closed", "backlog"],
                "in_progress": ["open", "blocked", "closed"],
                "blocked": ["in_progress", "closed"],
                "closed": ["open"],
            }
        },
        "initial_status": "open",
        "priorities": {
            0: {"name": "critical", "color": "red"},
            1: {"name": "high", "color": "bright_red"},
            2: {"name": "medium", "color": "yellow"},
            3: {"name": "low", "color": "blue"},
            4: {"name": "trivial", "color": "white"},
        },
        "default_priority": 2,
    }
    (root / ".kanbus.yml").write_text(
        yaml.safe_dump(payload, sort_keys=False), encoding="utf-8"
    )
    context.expected_project_dir = (root / "extras" / "project").resolve()


@given("a non-git directory without projects")
def given_non_git_directory(context: object) -> None:
    root = Path(context.temp_dir) / "no-git"
    root.mkdir(parents=True, exist_ok=True)
    context.working_directory = root


@given("a repository with a fake git root pointing to a file")
def given_repo_fake_git_root(context: object) -> None:
    root = _create_repo(context, "fake-git-root")
    fake_file = root / "not-a-dir"
    fake_file.write_text("data", encoding="utf-8")
    bin_dir = root / "fake-bin"
    bin_dir.mkdir(parents=True, exist_ok=True)
    git_path = bin_dir / "git"
    git_path.write_text(f"#!/bin/sh\necho {fake_file}\n", encoding="utf-8")
    git_path.chmod(0o755)
    context.original_path_env = os.environ.get("PATH", "")
    os.environ["PATH"] = f"{bin_dir}{os.pathsep}{context.original_path_env}"


@when("project directories are discovered")
def when_project_dirs_discovered(context: object) -> None:
    root = Path(context.working_directory)
    try:
        context.project_dirs = discover_project_directories(root)
        context.project_error = None
    except ProjectMarkerError as error:
        context.project_dirs = []
        context.project_error = str(error)


@when("kanbus configuration paths are discovered from the filesystem root")
def when_kanbus_configuration_paths_from_root(context: object) -> None:
    try:
        context.project_dirs = discover_kanbus_projects(Path("/"))
        context.project_error = None
    except ProjectMarkerError as error:
        context.project_dirs = []
        context.project_error = str(error)


@when("the project directory is loaded")
def when_project_dir_loaded(context: object) -> None:
    root = Path(context.working_directory)
    try:
        context.project_dir = load_project_directory(root)
        context.project_dirs = [context.project_dir]
        context.project_error = None
    except ProjectMarkerError as error:
        context.project_dir = None
        context.project_dirs = []
        context.project_error = str(error)


@when("the configuration path is requested")
def when_configuration_path_requested(context: object) -> None:
    root = Path(context.working_directory)
    try:
        context.configuration_path = get_configuration_path(root)
        context.project_error = None
    except ProjectMarkerError as error:
        context.configuration_path = None
        context.project_error = str(error)


@then("exactly one project directory should be returned")
def then_single_project_returned(context: object) -> None:
    assert len(context.project_dirs) == 1


@then('project discovery should fail with "project not initialized"')
def then_project_not_initialized(context: object) -> None:
    assert context.project_error == "project not initialized"


@then('project discovery should fail with "multiple projects found"')
def then_project_multiple(context: object) -> None:
    assert context.project_error is not None
    assert "multiple projects found" in context.project_error


@then('project discovery should fail with "kanbus path not found"')
def then_project_missing_path(context: object) -> None:
    assert "path not found" in context.project_error


@then('project discovery should fail with "unknown configuration fields"')
def then_project_unknown_fields(context: object) -> None:
    assert context.project_error == "unknown configuration fields"


@then("project discovery should include the referenced path")
def then_project_includes_referenced_path(context: object) -> None:
    expected = getattr(context, "expected_project_dir", None)
    assert expected is not None
    expected = expected.resolve()
    resolved = [path.resolve() for path in context.project_dirs]
    assert expected in resolved


@then("project discovery should return no projects")
def then_project_returns_no_projects(context: object) -> None:
    assert context.project_dirs == []


@then('configuration path lookup should fail with "project not initialized"')
def then_config_path_missing(context: object) -> None:
    assert context.project_error == "project not initialized"


@then('project discovery should fail with "Permission denied"')
def then_project_permission_denied(context: object) -> None:
    assert context.project_error is not None
    assert "Permission denied" in context.project_error
